src/TermInterface.jl

```diff
@@ -45,12 +45,41 @@ and pattern matching features.
 function exprhead end
 export exprhead
 
+"""
+  head(x)
+
+If `x` is a term as defined by `istree(x)`, `head(x)` returns the head of the
+term if `x`. The `head` type has to be provided by the package.
+"""
+function head end
+export head
+
+"""
+  head_symbol(x::HeadType)
+
+If `x` is a head object, `head_symbol(T, x)` returns a `Symbol` object that
+corresponds to `y.head` if `y` was the representation of the corresponding term
+as a Julia Expression. This is useful to define interoperability between
+symbolic term types defined in different packages and should be used when
+calling `maketerm`.
+"""
+function head_symbol end
+export head_symbol
+
+"""
+  tail(x)
+
+Get the arguments of `x`, must be defined if `istree(x)` is `true`.
+"""
+function tail end
+export tail
+
 
 """
   operation(x)
 
 If `x` is a term as defined by `istree(x)`, `operation(x)` returns the
-head of the term if `x` represents a function call, for example, the head
+operation of the term if `x` represents a function call, for example, the head
 is the function being called.
 """
 function operation end
```

Comment:
Again should not be in this package.

Reply:
While I believe this one should stay in this package.

Reply:
This package is used to define an AST interface so that pattern matchers can match against the AST and traverse it. I think it's a little bit of a mistake to introduce `@rule` with the example of `+(~a, ~b)`, because in many IRs we would want to write this as `:call(+, ~a, ~b)`. Not all ASTs will have something analogous to a function call, and it's unclear what benefit would be derived from standardizing this notion here. `operation` and `arguments` should live in packages that define ASTs that can give these functions meaning.

TL;DR: I don't think we can define what the new meaning of `operation` is supposed to do in a way that captures all of its possible use cases in downstream packages, so I think it should not be in this package.

Reply:
> Not all ASTs will have something analogous to a function call

I guess most of what are/used to be the dependents of this package do:
- Metatheory.jl `AbstractPat` AST for patterns 
- Julia Exprs
- SymbolicUtils IIRC terms have function calls and array indexing
- Everything depending on Symbolics
- Most other symbolic mathematics packages.